File: data-structure/map/map.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INITIAL_CAPACITY 10 // 初始容量

// Map 结构体
typedef struct
{
    char *key;
    int value;
} Entry;

typedef struct
{
    Entry *entries;
    int size;
    int capacity;
} Map;

// 初始化 Map
void initMap(Map *map)
{
    map->size = 0;
    map->capacity = INITIAL_CAPACITY;
    map->entries = (Entry *)malloc(map->capacity * sizeof(Entry));
}

// 重新分配容量
void resizeMap(Map *map, int newCapacity)
{
    map->capacity = newCapacity;
    map->entries = (Entry *)realloc(map->entries, map->capacity * sizeof(Entry));
}

// 插入键值对（如果存在则更新）
void put(Map *map, const char *key, int value)
{
    for (int i = 0; i < map->size; i++)
    {
        if (strcmp(map->entries[i].key, key) == 0)
        {
            map->entries[i].value = value; // 更新值
            return;
        }
    }
    if (map->size >= map->capacity)
    {
        resizeMap(map, map->capacity * 2);
    }
    map->entries[map->size].key = strdup(key);
    map->entries[map->size].value = value;
    map->size++;
}

// 查找键
int get(Map *map, const char *key, int *found)
{
    for (int i = 0; i < map->size; i++)
    {
        if (strcmp(map->entries[i].key, key) == 0)
        {
            *found = 1;
            return map->entries[i].value;
        }
    }
    *found = 0;
    return -1; // 未找到
}

// 删除键
void delete(Map *map, const char *key)
{
    for (int i = 0; i < map->size; i++)
    {
        if (strcmp(map->entries[i].key, key) == 0)
        {
            free(map->entries[i].key);
            for (int j = i; j < map->size - 1; j++)
            {
                map->entries[j] = map->entries[j + 1];
            }
            map->size--;
            return;
        }
    }
}

// 释放 Map
void freeMap(Map *map)
{
    for (int i = 0; i < map->size; i++)
    {
        free(map->entries[i].key);
    }
    free(map->entries);
}

// 测试
int main()
{
    Map map;
    initMap(&map);

    put(&map, "apple", 10);
    put(&map, "banana", 20);
    put(&map, "orange", 30);

    int found;
    printf("apple: %d\n", get(&map, "apple", &found));
    printf("banana: %d\n", get(&map, "banana", &found));
    printf("grape: %d (not found)\n", get(&map, "grape", &found));

    delete (&map, "banana");
    printf("banana after delete: %d\n", get(&map, "banana", &found));

    freeMap(&map);
    return 0;
}

/*
jarry@MacBook-Pro map % gcc map.c
jarry@MacBook-Pro map % ./a.out
apple: 10
banana: 20
grape: -1 (not found)
banana after delete: -1
*/

File: 01-data-structures/map/map.js
// JS基于数组模拟内置Map对象，为了演示数据结构
// 实际JS引擎采取使用 哈希表 + 链地址法 或用 红黑树
class Map {
    constructor() {
        this.entries = [];
        this.size = 0;
        this.capacity = 10;
    }

    // 重新分配容量，JS动态数组，无需调整容量
    resizeMap(newCapacity) {
        this.capacity = newCapacity;
    }

    // 插入键值对（如果存在则更新）
    put(key, value) {
        for (let i = 0; i < this.size; i++) {
            if (this.entries[i].key === key) {
                this.entries[i].value = value; // 更新值
                return;
            }
        }
        if (this.size >= this.capacity) {
            this.resizeMap(this.capacity * 2);
        }
        // 或push追加新键值对
        // this.entries.push({ key, value });
        this.entries[this.size] = {
            key,
            value
        };
        this.size++;
    }

    // 查找键
    get(key) {
        for (let i = 0; i < this.size; i++) {
            if (this.entries[i].key === key) {
                return this.entries[i].value;
            }
        }
        return -1; // 未找到
    }

    // 删除键
    delete(key) {
        for (let i = 0; i < this.size; i++) {
            if (this.entries[i].key === key) {
                this.entries.splice(i, 1);
                this.size--;
                return;
            }
        }
    }

    // 判断键是否存在
    has(key) {
        return this.entries.some(entry => entry.key === key);
    }

    // 获取键值对个数
    getSize() {
        return this.entries.length;
    }

    // 清空所有数据
    clear() {
        this.entries = [];
    }

    // 遍历所有键值对
    forEach(callback) {
        this.entries.forEach(({
            key,
            value
        }) => callback(value, key, this));
    }
}

// 测试
const map = new Map();

map.put("apple", 10);
map.put("banana", 20);
map.put("orange", 30);

console.log("apple:", map.get("apple"));
console.log("banana:", map.get("banana"));
console.log("grape:", map.get("grape"));

map.delete("banana");
console.log("banana after delete:", map.get("banana"));

/*
jarry@MacBook-Pro map %  node map.js 
apple: 10
banana: 20
grape: -1
banana after delete: -1
*/